src/gc/impl/manual/gc.d

```diff
@@ -0,0 +1,305 @@
+        }
+    }
+
+    this()
+    {
+    }
+
+    void Dtor()
+    {
+        cstdlib.free(roots);
+        cstdlib.free(ranges);
+    }
+
+    void enable()
+    {
+    }
+
+    void disable()
+    {
+    }
+
+    void collect() nothrow
+    {
+    }
+
+    void collectNoStack() nothrow
+    {
+    }
+
+    void minimize() nothrow
+    {
+    }
+
+    uint getAttr(void* p) nothrow
+    {
+        return 0;
+    }
+
+    uint setAttr(void* p, uint mask) nothrow
+    {
+        return 0;
+    }
+
+    uint clrAttr(void* p, uint mask) nothrow
+    {
+        return 0;
+    }
+
+    void* malloc(size_t size, uint bits, const TypeInfo ti) nothrow
+    {
+        void* p = cstdlib.malloc(size);
+
+        if (size && p is null)
+            onOutOfMemoryError();
+        return p;
+    }
+
+    BlkInfo qalloc(size_t size, uint bits, const TypeInfo ti) nothrow
+    {
+        BlkInfo retval;
+        retval.base = malloc(size, bits, ti);
+        retval.size = size;
+        retval.attr = bits;
+        return retval;
+    }
+
+    void* calloc(size_t size, uint bits, const TypeInfo ti) nothrow
+    {
+        void* p = cstdlib.calloc(1, size);
+
+        if (size && p is null)
+            onOutOfMemoryError();
+        return p;
+    }
+
+    void* realloc(void* p, size_t size, uint bits, const TypeInfo ti) nothrow
+    {
+        p = cstdlib.realloc(p, size);
+
+        if (size && p is null)
+            onOutOfMemoryError();
+        return p;
+    }
+
+    size_t extend(void* p, size_t minsize, size_t maxsize, const TypeInfo ti) nothrow
+    {
+        return 0;
+    }
+
+    size_t reserve(size_t size) nothrow
+    {
+        return 0;
+    }
+
+    void free(void* p) nothrow
+    {
+        cstdlib.free(p);
+    }
+
+    /**
+     * Determine the base address of the block containing p.  If p is not a gc
+     * allocated pointer, return null.
+     */
+    void* addrOf(void* p) nothrow
+    {
+        return null;
+    }
+
+    /**
+     * Determine the allocated size of pointer p.  If p is an interior pointer
+     * or not a gc allocated pointer, return 0.
+     */
+    size_t sizeOf(void* p) nothrow
+    {
+        return 0;
+    }
+
+    /**
+     * Determine the base address of the block containing p.  If p is not a gc
+     * allocated pointer, return null.
+     */
+    BlkInfo query(void* p) nothrow
+    {
+        return BlkInfo.init;
+    }
+
+    GCStats stats() nothrow
+    {
+        return GCStats.init;
+    }
+
+    void addRoot(void* p) nothrow @nogc
+    {
+        Root* r = cast(Root*) cstdlib.realloc(roots, (nroots + 1) * roots[0].sizeof);
+        if (r is null)
+            onOutOfMemoryError();
+        r[nroots++] = p;
+        roots = r;
+    }
+
+    void removeRoot(void* p) nothrow @nogc
+    {
+        for (size_t i = 0; i < nroots; ++i)
+        {
+            if (roots[i] is p)
+            {
+                roots[i] = roots[--nroots];
+                return;
+            }
+        }
```

Comment:
Is this correct? Don't you need to move all of the elements after and including the removed one to the left? I think the correct way should be something along the lines:

```
size_t idx;
for (; roots[idx] !is p && idx < nroots; idx++) { }
assert (roots[idx] is p);
roots[idx .. nroots - 1] = roots[idx + 1 .. nroots]; // move elemnts to the left
```